src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/EditingTargetCodeExecution.java

```diff
@@ -158,6 +161,44 @@ public void executeRange(Range range)
       executeRange(range, null, false);
    }
    
+   public void sendSelectionToTerminal(boolean terminalExecuteWhenNotFocused)
+   {
+      if (terminalExecuteWhenNotFocused && !docDisplay_.isFocused())
+      {
+         // the current editor may be in another window; if one of our source
+         // windows was last focused, use that one instead
+         SourceWindowManager manager = 
```

Comment:
I think you can simply this (and remove the dummy `WithoutFocus` variant of the command) by using a cross-window event. If you fire one of those, it proceeds as normal in the main window, but in a satellite window it will fire the event to the main window for processing instead.

You can even do this in an auxiliary mode so that the main window is raised enough to show the result but not enough to obscure the source window. See this class for more: https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/application/events/CrossWindowEvent.java